frugen.c

```diff
@@ -460,6 +530,101 @@ int main(int argc, char *argv[])
 					fatal("JSON support was disabled at compile time");
 #endif
 				}
+				else if (use_binary) {
+					int fd = open(optarg, O_RDONLY);
+					if (fd < 0) {
+						perror("");
+						fatal("Failed to open file");
+					}
+
+					char common_header[8];
+					safe_read(fd, common_header, 8);
+
+					// Common Header Format Version = common_header[0]
+					// Internal Use Area Starting Offset = common_header[1]
+					uint16_t chassis_start_offset = 8 * common_header[2];
+					uint16_t board_start_offset = 8 * common_header[3];
+					uint16_t product_start_offset = 8 * common_header[4];
+					// MultiRecord Area Starting Offset = 8 * common_header[5]
+					// Padding = common_header[6] = 0
+					// Checksum = common_header[7]
+
+					// For the above offsets, an offset of 0 is valid and implies
+					// that the field is not present.
+
+					bool data_has_chassis = chassis_start_offset != 0;
+					bool data_has_board = board_start_offset != 0;
+					bool data_has_product = product_start_offset != 0;
+
+					if (data_has_chassis) {
+						lseek(fd, chassis_start_offset, SEEK_SET);
+
+						uint8_t chassis_header[3];
+						safe_read(fd, chassis_header, 3);
+						if (chassis_header[0] != 1)
+							fatal("Unsupported Chassis Info Area Format Version");
+						// Chassis Info Area Length = 8 * chassis_header[1]
+						chassis.type = chassis_header[2];
+						fd_read_field(fd, chassis.pn);
+						fd_read_field(fd, chassis.serial);
+						fd_fill_custom_fields(fd, &chassis.cust);
+
+						has_chassis = true;
+					}
+					if (data_has_board) {
+						lseek(fd, board_start_offset, SEEK_SET);
+
+						uint8_t board_header[3];
+						safe_read(fd, board_header, 3);
+						if (board_header[0] != 1)
+							fatal("Unsupported Board Info Area Format Version");
+						// Board Info Area Length = 8 * board_header[1]
+						board.lang = board_header[2];
+
+						uint32_t min_since_1996 = 0;
+						safe_read(fd, &min_since_1996, 3);
+						struct tm tm_1996 = {
+							.tm_year = 96,
+							.tm_mon = 0,
+							.tm_mday = 1
+						};
+						// The argument to mktime is zoneless
+						board.tv.tv_sec = mktime(&tm_1996) + 60 * min_since_1996;
```

Comment:
There is already a very similar piece of code in `fru.c`, please think about some function that would allow to get rid of repetitions.